### PR TITLE
feat: add Prometheus usage beacon module

### PR DESCRIPTION
**Pull request by iliasabk**
## Summary

Implements the Prometheus Protocol Rust usage beacon for the open app bounty #4.

- Adds public Candid-compatible CallerActivity and UsageStats payload types.
- Adds init, track_call, and start_timer APIs.
- Aggregates calls by caller principal and MCP tool, then reports them through log_call.
- Uses ic-cdk-timers for recurring reporting and restarts timers safely.
- Restores a batch when the tracker rejects or cannot decode a report, so usage is not silently lost.
- Adds unit coverage for aggregation, exact payload construction, timestamp advancement, and failed-delivery retry state.

## Validation

The repository CI will run formatting, clippy, docs, native tests, and the wasm build. Rust is not installed in the local execution environment, so I have not claimed local cargo results.

Related to Prometheus Protocol app bounty #4.

### CODE REVIEW

**Comment by iliasaberkane6-lab** [comment]
The initial CI workflow is marked action_required because this is a first-time fork PR; there are no job results yet. Please approve the workflow run when convenient so the repository checks can validate formatting, clippy, tests, docs, and the wasm build.

**Comment by iliasaberkane6-lab** [comment]
Validation update: the fork CI run completed successfully for the implementation (fmt, docs, native tests, wasm build, and clippy). The first run only found clone-on-Copy Clippy diagnostics for Principal; those were fixed in commit 55ff7f1. Successful validation run: https://github.com/iliasaberkane6-lab/ic-rmcp/actions/runs/34753938023

File: ic-rmcp/src/beacon.rs [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+
+/// One caller's usage of one MCP tool during a reporting window.
+#[derive(Clone, Debug, CandidType, Deserialize, Eq, PartialEq)]
+pub struct CallerActivity {
+    /// Principal that invoked the tool.
+    pub caller: Principal,
+    /// Application-defined MCP tool name or identifier.
+    pub tool_id: String,
+    /// Number of calls made by `caller` to `tool_id`.
+    pub call_count: u64,
```
Comment (okoyfoeciov) [inline ic-rmcp/src/beacon.rs]: `call_count: u64` and `UsageStats.{start,end}_timestamp_ns: i64` (line 37/39) encode as Candid `nat64`/`int64`. The real UsageTracker canister (prometheus-protocol/motoko-sdk `canisters/usage_tracker/src/main.mo`) declares these as Motoko `Nat`/`Int` — unbounded `nat`/`int` on the wire, with no auto-coercion from fixed-width types in Candid. Every `log_call` to the real tracker will fail to decode. Use `candid::Nat`/`candid::Int` (or u128/i128) instead, and add a round-trip encode/decode test — none of the current tests exercise real Candid serialization, which is why this wasn't caught.

File: ic-rmcp/src/beacon.rs [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    context.state.borrow_mut().last_send_timestamp_ns = ic_cdk::api::time() as i64;
+
+    let state = Rc::clone(&context.state);
+    let tracker_canister_id = context.tracker_canister_id;
+    let interval = Duration::from_secs(context.reporting_interval_seconds);
+    context.timer_id = Some(set_timer_interval(interval, move || {
+        let state = Rc::clone(&state);
+        let tracker_canister_id = tracker_canister_id;
+        ic_cdk::futures::spawn(async move {
+            let _ = send_beacon(state, tracker_canister_id).await;
```
Comment (okoyfoeciov) [inline ic-rmcp/src/beacon.rs]: Send failures are silently discarded here, and a failed send restores the counted data (`restore_usage`) rather than dropping it. Combined with the type mismatch above, `usage_data` will never actually clear against the real tracker and grows unboundedly with no visibility. At minimum, log the error.

File: ic-rmcp/src/beacon.rs [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+///
+/// Restarting first cancels the previous timer, which makes this safe to call
+/// after canister initialization code is re-run. The callback never holds a
+/// `RefCell` borrow while awaiting the tracker call.
+pub fn start_timer(context: &mut BeaconContext) {
+    if let Some(timer_id) = context.timer_id.take() {
+        clear_timer(timer_id);
+    }
+
+    context.state.borrow_mut().last_send_timestamp_ns = ic_cdk::api::time() as i64;
```
Comment (okoyfoeciov) [inline ic-rmcp/src/beacon.rs]: This unconditionally resets `last_send_timestamp_ns` to `now` on every `start_timer()` call, even if unreported usage already exists, understating the next window's duration. Overlapping `send_beacon` tasks (the interval floor is 1s) can also commit this out of order. Consider only setting it when still at its default, or using a monotonic `max()`.

File: ic-rmcp/Cargo.toml [hunk 4]
```diff
@@ -7,15 +7,16 @@ license = "MIT"
 authors = ["Kyle <lhkhiem@outlook.com>", "Long Tran <tranbaolong14@gmail.com>"]
 
 [dependencies]
+candid = "0.10"
 serde = { version = "1.0", features = ["derive", "rc"] }
 serde_json = "1.0"
 ic-http-certification = "3.0.3"
 rmcp = "0.2.1"
 ic-cdk = { version = "0.18", features = ["transform-closure"] }
+ic-cdk-timers = "0.12.1"
```
Comment (okoyfoeciov) [inline ic-rmcp/Cargo.toml]: `candid` and `ic-cdk-timers` are added as mandatory deps and `beacon` is compiled in unconditionally, but the feature is opt-in at the API level. Consider gating both behind a `beacon` Cargo feature so non-beacon consumers don't pay the extra compile/wasm-size cost.

**Review by okoyfoeciov — CHANGES_REQUESTED** [review]
Checked this against the real target: Prometheus Protocol's bounty #4 and the actual UsageTracker canister (prometheus-protocol/motoko-sdk) are both legitimate and this PR is correctly scoped to them. However, there's a blocking wire-format bug plus a couple of correctness/design issues below — requesting changes before merge.

**Comment by iliasaberkane6-lab** [comment]
Addressed the review feedback in commit 0f61b75:\n\n- encoded UsageStats timestamps as candid::Int and call counts as candid::Nat to match the Motoko Int/Nat schema\n- added a Candid encode/decode round-trip test\n- made report timestamps monotonic and logged timer delivery failures\n- gated the beacon module, candid, and ic-cdk-timers behind the beacon feature\n\nValidation: cargo fmt --all -- --check; cargo test -p ic-rmcp --features beacon (11 unit + 5 integration tests); cargo test -p ic-rmcp (7 unit + 5 integration tests); cargo clippy -p ic-rmcp --features beacon --all-targets -- -D warnings.

**Comment by iliasaberkane6-lab** [comment]
Quick question @okoyfoeciov — the bounty docs ask to register submissions in the Discord #bounties / 💰-reserve-bounty channel, but I joined the server and can't find either channel. Is Discord still the right place to register these PRs (#32 for bounty #4, #33 for bounty #1) and coordinate payout, or can we confirm it here? Thanks!

**Comment by okoyfoeciov** [comment]
Hi @iliasaberkane6-lab, we are not Prometheus. We used to work with Prometheus owner. But we are not them. I also can't find the bounty information in the Discord channel. 
All maintainers in this repo are busy with other projects. So we have to close your PRs soon (#32, #33, #34, #35). Feel free to fork the repo and do your own change. Thank you!

**Comment by iliasaberkane6-lab** [comment]
Thanks for clarifying, @okoyfoeciov — and for the thorough reviews; the wire-format and trust-boundary feedback was genuinely useful.

Understood on closing the PRs. For context: these implement Prometheus Protocol's published bounties (#4 beacon, #1 API key, #2 treasury) from prometheus-protocol/app-bounties, which reference ic-rmcp as the target repo. Since the sponsor is Prometheus rather than this team, I'll move coordination to the sponsor's channels — I haven't been able to locate their Discord either, so if you happen to have a direct contact for the Prometheus team, a pointer would be appreciated.

The implementations stay available on my fork branches in case the sponsor wants to pick them up. Thanks again for the time you put into reviewing!
